### PR TITLE
Lensing

### PR DESCRIPTION
**Pull request by Maxelee**
I started to mess around with converting lensing functions in vmad/MADLens to tf. I added to flowpm/lenspm.py: 

* generate: Creates the rotation matrices and box shifts to fill the light cone. It finds the num of rotations and shifts needed between a fast/flowpm step and returns a list of them. It only allows for 90 deg box rotations (6 unique rotations) and two box replications (in x-y plane)

* rotate: applies the rotation matrices to a 3D tensor. This is really just an einsum between the particle positions (modulo box size) and an additional shift in the z-direction. If there is a box replication needed in the x, y, direction, it performs this shift as well. 

* wlen: finds the weights for the Lensing efficiency. based on the mesh that we're projecting too, and the mesh we're simulating we can find the column density of particles of the light cone. We also need to find the redshift of the particles and then can compute the weights... I think I did this right, but am not 100% sure if I'm using tf.RegisterGradient correctly

* makemap: I started working on this but got stuck. The painting process in vmad/fastpm/pmesh I'm not sure if there's a good way to convert that all to tensorflow. 

I made a change in utils.py as well in c2r and r2c to allow for 2D fields and do a 2D fft which we need for projecting. I'm not sure if the syntax I used is the best way to do it, but I put it in. I also copied the 3D test case then changed the test field to 2D in tests/test_utils.py

I made a notebook showing all of the functions and some examples of how they work to make sure its working right, I also started messing around with auto diff in tf to make sure I understand the syntax. notebooks/Lensing.ipynb

### CODE REVIEW

**Comment by EiffL** [comment]
Amazing @Maxelee :-D I'm having a look right now

File: flowpm/lenstf.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        return xy, d
+
+
+
+def z_chi(d, cosmo, name='z_chi'):
+    with tf.name_scope(name):
+        # redshift as a function of comsoving distance for underlying cosmology
+        z_int          = np.logspace(-12,np.log10(1500),40000)
+        chis           = cosmo.comoving_distance(z_int) #Mpc/h
+        z_chi_int = scipy.interpolate.interp1d(chis, z_int, kind='linear', bounds_error=False, fill_value='extrapolate')
```
Comment (EiffL) [inline flowpm/lenstf.py:74]: we can probably use a TensorFlow method for this. I know at least of the TensorFlow Probability interp_regular_1d_grid: https://www.tensorflow.org/probability/api_docs/python/tfp/math/interp_regular_1d_grid
This way the function would be directly autodiffable

File: flowpm/utils.py [hunk 2]
```diff
@@ -115,7 +115,7 @@ def cic_readout(mesh, part, name="CiCReadout"):
     value = tf.reduce_sum(weightedvals, axis=-1)
     return value
 
-def r2c3d(rfield, norm=None, dtype=tf.complex64, name="R2C3D"):
+def r2c3d(rfield, norm=None, dims=3, dtype=tf.complex64, name="R2C3D"):
```
Comment (EiffL) [inline flowpm/utils.py:118]: we probably should have a new function named `r2c2d` to handle the 2d case, just like tf.signal has both `fft3d` and `fft2d`

**Review by EiffL — CHANGES_REQUESTED** [review]
Looking good :-) I've made some minor comments to use TF ops for interpolation

File: flowpm/lenstf.py (hunk 1 above)
Comment (Maxelee) [reply flowpm/lenstf.py:74]: Awesome! ok Ill get that implemented

File: flowpm/utils.py (hunk 2 above)
Comment (Maxelee) [reply flowpm/utils.py:118]: Ok cool. I wasn't sure which way would be better, but that makes sense. I'll have that changed!

**Comment by EiffL** [comment]
Let me go through your notebook and I may have more comments :-)

**Comment by EiffL** [comment]
Okkkk I see I see, very nice notebook :-) 
So, if I understand correctly, you only need a TensorFlow 2d painting tool. I can write you that :-) The 3D version of cic painting is here: https://github.com/modichirag/flowpm/blob/a1574363e0ba14f091c3905ab85d39147553a9bb/flowpm/utils.py#L9
It's only a matter of removing one dimension and adjusting for the units of your xy coordinates. I'm gonna try to add this

**Comment by EiffL** [comment]
Ok, so, I've added a `cic_paint_2d` function which works like this:
```
from flowpm.utils import cic_paint_2d
import tensorflow as tf

# Create an empty mesh
mesh = tf.zeros([1,32,32])
# Sample some particles in mesh coordinates
parts = 2*tf.random.normal((1,32*32, 2), name='particle_pos') + 16
# Paint
im = cic_paint_2d(mesh,parts)
# Profit
plt.imshow(im.numpy()[0])
```

Note that to make things simple, it's simpler if the particles are in "mesh coordinates" when you paint them. So for a 32x32 mesh, you want your particles to have x,y position between 0 and 32

**Review by EiffL — APPROVED** [review]
Thanks Max, let's merge this into the code base on the lensing branch and then we can continue to contribute to it together

**Pull request merged**